netcore/System.Private.CoreLib/src/System.Threading/LowLevelLifoSemaphore.Unix.cs

```diff
@@ -0,0 +1,131 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using Internal.Runtime.CompilerServices;
+
+namespace System.Threading
+{
+	internal unsafe sealed partial class LowLevelLifoSemaphore : IDisposable
+	{
+		struct WaitEntry
+		{
+			public object condition;
+			public bool signaled;
+			public void* previous;
+			public void* next;
+		}
+
+		private object mutex;
+		private void* head;
+		private uint pending_signals;
+
+		private void Create (int maximumSignalCount)
+		{
+			mutex = new object();
+			head = null;
+			pending_signals = 0;
+		}
+
+		public void Dispose ()
+		{
+		}
+
+		private bool WaitCore (int timeoutMs)
+		{
+			WaitEntry wait_entry = new WaitEntry ();
+			bool mutexLocked = false;
+			bool waitEntryLocked = false;
+
+			try {
+				Monitor.try_enter_with_atomic_var (mutex, Timeout.Infinite, false, ref mutexLocked);
+
+				if (pending_signals > 0) {
+					--pending_signals;
+					return true;
+				}
+
+				wait_entry.condition = new object();
```

Comment:
Wait requires a heap allocation? Ouch.

Reply:
I wish I could write it in some other way. In native code I could use `pthread_condition_t` on a stack...

Reply:
...and it was also cleaner because I could use just one `pthread_mutex_t` for all the condition variables.

Reply:
Maybe it should be native then?

Reply:
Or some unusual intrinsics?

Reply:
Here's the same thing in native code for reference: https://github.com/filipnavara/corefx/blob/5c9efb5efa42770cfbb2bffad59fdb988d21a0c2/src/Native/Unix/System.Native/pal_threading.c#L168-L280

Reply:
So this:

https://github.com/filipnavara/corefx/blob/5c9efb5efa42770cfbb2bffad59fdb988d21a0c2/src/Native/Unix/System.Native/pal_threading.c#L199

may or may not heap allocate.
Probably not on Linux, probably on Cygwin (crazy), probably not on Windows (which has basically the same construct). I'd have to check Mac.

Maybe could kinda intrinsify the size/allocation and static initialization of stuff like Posix mutex and semaphore?
At least size, is more important and eaiser. Could still do a runtime initialization. JIT would just say sizeof (x) and not really be platform specific source.

Could seriously, build that as a separate layer, and then reuse it here.
So C# programmers could directly use pthread mutex/semaphore, with same/similar perf as C++, i.e. w/o extra allocations.

Reply:
And in fact, this would potentially allow lifting more code out of native and into managed. Anything using the native synchronization constructs.
On Windows btw it is very easy -- the native srwlock and condvar, are sizeof(pointer), and static initialization is zero-ing. We could only be so lucky on Posix.

Reply:
On Windows the managed code uses P/Invoke to I/O completion queue API to achieve the same thing. This is purely for Unix systems. My original plan was to push this piece of code to System.Native but there was some resistance to it on the basis that CoreRT already has a working implementation in managed code (the whole WaitSubsystem responsible for Mutex, Semaphore, etc. implementation) and if Mono wants something else it should be done in Mono.

I can slightly improve the situation with allocation by keeping the objects reusable using thread-static or in an allocation pool.

The reason I didn't port the native code to Mono is that the icalls would be Unix only. Not sure if that's a good idea.

Reply:
How good or bad is this idea?

 icall to get size of pthread_condvar_t.
 Store the size in .cctor.
 stackalloc it
 icall to init it
 icall to wait on it

Optionally:
 icall to query if static init is zero-ing
 and then skip icall to init if so (but zero it)

Reply:
Unix-only icalls are ok.

Reply:
Why not init and wait in one icall? that would be good enough for this usecase